README.md
# zapier-hubspot-additional
Advanced HubSot integration to do more complicated things than the standard HubSpot integration. For example, the standard HubSpot integration's find ticket action only allows for very simple ticket search filters. This integration exposes the filter parameter so that you can use the full range of filter flexibility, and thus use fewer tasks. 

Designed for authentication as a private app in HubSpot speak.


CHANGELOG.md
## 1.0.6

* Add logging for submission and response

## 1.0.5

* Add logging for ticket ID input

## 1.0.4

* Add handling of ISO date time string as a time input format

## 1.0.3

* Add SMS event handling

## 1.0.2

* Correctly returning multiple items as line items

## 1.0.1

* Added state filter to only return tickets in desired states

## 1.0.0

First version with working ticket retrieval


test/creates/sms_event.js
require('should');

const zapier = require('zapier-platform-core');

const App = require('../../index');
const appTester = zapier.createAppTester(App);

describe('Create - log SMS received using tickets retrieved by search and Date object', () => {
  zapier.tools.env.inject();

  it('should get a modified count of 2', async () => {
    const bundle = {
      authData: {
        hapi_key: process.env.HUBSPOT_API_KEY,
      },

      inputData: {
			tickets: [{"id":"161391783","properties":{"content":"You can ignore this","createdate":"2020-07-16T14:05:16.882Z","hs_lastmodifieddate":"2021-11-25T11:03:34.136Z","hs_object_id":"161391783","hs_pipeline":"6098385","hs_pipeline_stage":"6098389","hs_ticket_category":null,"hs_ticket_priority":"LOW","subject":"Test 3"},"createdAt":"2020-07-16T14:05:16.882Z","updatedAt":"2021-11-25T11:03:34.136Z","archived":false},{"id":"161521798","properties":{"content":"You can ignore this","createdate":"2020-07-16T13:50:11.754Z","hs_lastmodifieddate":"2021-11-25T11:03:34.690Z","hs_object_id":"161521798","hs_pipeline":"0","hs_pipeline_stage":"4","hs_ticket_category":null,"hs_ticket_priority":"LOW","subject":"Test 2"},"createdAt":"2020-07-16T13:50:11.754Z","updatedAt":"2021-11-25T11:03:34.690Z","archived":false}],
			from_phone: '07815142001',
			to_phone: '07451280992',
			event_type: 'received',
			message: 'Received: Test message',
			time: Date.now(),
		},
    };

    const response = await appTester(
      App.creates['sms_event'].operation.perform,
      bundle
    );
    // console.log(response[0]);
    response.modified_count.should.equal(2);
  });
});

describe('Create - log SMS received using tickets retrieved by search and ISO date', () => {
  zapier.tools.env.inject();

  it('should get a modified count of 2', async () => {
    const bundle = {
      authData: {
        hapi_key: process.env.HUBSPOT_API_KEY,
      },

      inputData: {
			tickets: [{"id":"161391783","properties":{"content":"You can ignore this","createdate":"2020-07-16T14:05:16.882Z","hs_lastmodifieddate":"2021-11-25T11:03:34.136Z","hs_object_id":"161391783","hs_pipeline":"6098385","hs_pipeline_stage":"6098389","hs_ticket_category":null,"hs_ticket_priority":"LOW","subject":"Test 3"},"createdAt":"2020-07-16T14:05:16.882Z","updatedAt":"2021-11-25T11:03:34.136Z","archived":false},{"id":"161521798","properties":{"content":"You can ignore this","createdate":"2020-07-16T13:50:11.754Z","hs_lastmodifieddate":"2021-11-25T11:03:34.690Z","hs_object_id":"161521798","hs_pipeline":"0","hs_pipeline_stage":"4","hs_ticket_category":null,"hs_ticket_priority":"LOW","subject":"Test 2"},"createdAt":"2020-07-16T13:50:11.754Z","updatedAt":"2021-11-25T11:03:34.690Z","archived":false}],
			from_phone: '07815142001',
			to_phone: '07451280992',
			event_type: 'received',
			message: 'Received: Test message',
			time: '2021-11-11T13:26:03.697Z',
		},
    };

    const response = await appTester(
      App.creates['sms_event'].operation.perform,
      bundle
    );
    // console.log(response[0]);
    response.modified_count.should.equal(2);
  });
});

describe('Create - log SMS sent successfully using tickets retrieved by search', () => {
  zapier.tools.env.inject();

  it('should get a modified count of 2', async () => {
    const bundle = {
      authData: {
        hapi_key: process.env.HUBSPOT_API_KEY,
      },

      inputData: {
			tickets: [{"id":"161391783","properties":{"content":"You can ignore this","createdate":"2020-07-16T14:05:16.882Z","hs_lastmodifieddate":"2021-11-25T11:03:34.136Z","hs_object_id":"161391783","hs_pipeline":"6098385","hs_pipeline_stage":"6098389","hs_ticket_category":null,"hs_ticket_priority":"LOW","subject":"Test 3"},"createdAt":"2020-07-16T14:05:16.882Z","updatedAt":"2021-11-25T11:03:34.136Z","archived":false},{"id":"161521798","properties":{"content":"You can ignore this","createdate":"2020-07-16T13:50:11.754Z","hs_lastmodifieddate":"2021-11-25T11:03:34.690Z","hs_object_id":"161521798","hs_pipeline":"0","hs_pipeline_stage":"4","hs_ticket_category":null,"hs_ticket_priority":"LOW","subject":"Test 2"},"createdAt":"2020-07-16T13:50:11.754Z","updatedAt":"2021-11-25T11:03:34.690Z","archived":false}],
			from_phone: '07815142001',
			to_phone: '07451280992',
			event_type: 'delivery_success',
			message: 'Sent: Test message',
			time: Date.now(),
		},
    };

    const response = await appTester(
      App.creates['sms_event'].operation.perform,
      bundle
    );
    // console.log(response[0]);
    response.modified_count.should.equal(2);
  });
});

describe('Create - log SMS not sent successfully using tickets retrieved by search with UTC string', () => {
  zapier.tools.env.inject();

  it('should get a modified count of 2', async () => {
    const bundle = {
      authData: {
        hapi_key: process.env.HUBSPOT_API_KEY,
      },

      inputData: {
			tickets: [{"id":"161391783","properties":{"content":"You can ignore this","createdate":"2020-07-16T14:05:16.882Z","hs_lastmodifieddate":"2021-11-25T11:03:34.136Z","hs_object_id":"161391783","hs_pipeline":"6098385","hs_pipeline_stage":"6098389","hs_ticket_category":null,"hs_ticket_priority":"LOW","subject":"Test 3"},"createdAt":"2020-07-16T14:05:16.882Z","updatedAt":"2021-11-25T11:03:34.136Z","archived":false},{"id":"161521798","properties":{"content":"You can ignore this","createdate":"2020-07-16T13:50:11.754Z","hs_lastmodifieddate":"2021-11-25T11:03:34.690Z","hs_object_id":"161521798","hs_pipeline":"0","hs_pipeline_stage":"4","hs_ticket_category":null,"hs_ticket_priority":"LOW","subject":"Test 2"},"createdAt":"2020-07-16T13:50:11.754Z","updatedAt":"2021-11-25T11:03:34.690Z","archived":false}],
			from_phone: '07815142001',
			to_phone: '07451280992',
			event_type: 'delivery_failure',
			message: 'Failed: Test message',
			time: Date.now(),
		},
    };

    const response = await appTester(
      App.creates['sms_event'].operation.perform,
      bundle
    );
    // console.log(response[0]);
    response.modified_count.should.equal(2);
  });
});


test/searches/find_tickets.js
require('should');

const zapier = require('zapier-platform-core');

const App = require('../../index');
const appTester = zapier.createAppTester(App);

describe('Search - find_tickets no state filter', () => {
  zapier.tools.env.inject();

  it('should get an array with more than 1 element', async () => {
    const bundle = {
      authData: {
        access_token: process.env.ACCESS_TOKEN,
      },

      inputData: {
			filter: '[{"filters":[{"value":"TESTVALUE","propertyName":"vehicle","operator":"EQ"}]}]'
		},
    };

    const response = await appTester(
      App.searches['find_tickets'].operation.perform,
      bundle
    );
    //~ console.log(response[0]);
    response[0].results.should.be.an.Array();
    response[0].results.length.should.be.aboveOrEqual(1);
    response[0].count.should.equal(response[0].results.length);
    response[0].results[0].should.have.property('id');
  });
  
  
});

describe('Search - find_tickets with state filter', () => {
  zapier.tools.env.inject();

  it('should get an array with 1 element containg ticket 161521798', async () => {
    const bundle = {
      authData: {
        access_token: process.env.ACCESS_TOKEN,
      },

      inputData: {
			filter: '[{"filters":[{"value":"TESTVALUE","propertyName":"vehicle","operator":"EQ"}]}]',
			state_inclusion: {
				4: 1
			}
		},
    };

    const response = await appTester(
      App.searches['find_tickets'].operation.perform,
      bundle
    );
    response[0].results.should.be.an.Array();
    response[0].results.length.should.equal(1);
    response[0].count.should.equal(response[0].results.length);
    response[0].results[0].id.should.equal('161521798');
    response[0].results[0].properties.hs_pipeline_stage.should.equal('4');
  });
});

describe('Search - find_tickets with complex state filter', () => {
  zapier.tools.env.inject();

  it('should get an array with 2 elements', async () => {
    const bundle = {
      authData: {
        access_token: process.env.ACCESS_TOKEN,
      },

      inputData: {
			filter: '[{"filters":[{"value":"TESTVALUE","propertyName":"vehicle","operator":"EQ"}]}]',
			state_inclusion: {
				6098389: 1,
				4: 1
			}
		},
    };

    const response = await appTester(
      App.searches['find_tickets'].operation.perform,
      bundle
    );
    //~ console.log(response[0]);
    response[0].results.should.be.an.Array();
    response[0].results.length.should.be.aboveOrEqual(2);
    response[0].count.should.equal(response[0].results.length);
    console.log("Ticket 0: " + response[0].results[0].id + ", state " + response[0].results[0].properties.hs_pipeline_stage);
    console.log("Ticket 1: " + response[0].results[1].id + ", state " + response[0].results[0].properties.hs_pipeline_stage);
    response[0].results[0].id.should.equal('161391783');
    response[0].results[1].id.should.equal('161521798');
    response[0].results[1].properties.hs_pipeline_stage.should.equal('4');
  });
});

describe('Search - find_tickets with non-matching state filter', () => {
  zapier.tools.env.inject();

  it('should be an empty array', async () => {
    const bundle = {
      authData: {
        access_token: process.env.ACCESS_TOKEN,
      },

      inputData: {
			filter: '[{"filters":[{"value":"TESTVALUE","propertyName":"vehicle","operator":"EQ"}]}]',
			state_inclusion: {
				999999: 1
			}
		},
    };

    const response = await appTester(
      App.searches['find_tickets'].operation.perform,
      bundle
    );
    response[0].results.should.be.an.Array();
    response[0].results.length.should.equal(0);
    response[0].count.should.equal(response[0].results.length);
    console.log(response[0]);
  });
});


authentication.js
module.exports = {
  type: 'custom',
  connectionLabel: '{{hub_id}}',
  test: {
    url: 'https://api.hubapi.com/crm/v3/objects/tickets?limit=10&archived=false',
    method: 'GET',
    params: {},
    headers: { Authorization: 'Bearer {{bundle.authData.access_token}}' },
    body: {},
    removeMissingValuesFrom: {},
  },
  fields: [
    {
      computed: false,
      key: 'hub_id',
      required: true,
      label: 'Hub ID',
      type: 'string',
      helpText: 'The ID of the Hub you want to connect to. First number in the URL you are redirected to after going to https://app.hubspot.com',
    },
    {
      computed: false,
      key: 'access_token',
      required: true,
      label: 'Private app access token (needed for access to CRM API)',
      type: 'password',
      helpText: 'Go to HubSpot > Settings > Integrations > Private Apps - https://app.hubspot.com/private-apps/<yourhubID>. Scopes required: tickets, crm.ojects.contacts.read',
    },
    {
      computed: false,
      key: 'hapi_key',
      required: true,
      label: 'Full access API key (needed for access to Engagements API)',
      type: 'password',
      helpText: 'Go to HubSpot > Settings > Integrations > API Key - https://app.hubspot.com/api-key/<yourhubID>/call-log',
    },
  ],
  customConfig: {},
};


creates/sms_event.js
const perform = async (z, bundle) => {
	var engagement = {
		engagement: {
			active: true,
			ownerId: 851,
			type: "CALL",
		},
		associations: {
		},
		metadata: {
			toNumber: bundle.inputData.to_phone,
			fromNumber: bundle.inputData.from_phone,
			durationMilliseconds: 1000,
			body: bundle.inputData.message
		}
	}
	console.log(bundle.inputData.tickets[0]);
	const options = {
	url: 'https://api.hubapi.com/engagements/v1/engagements',
	method: 'POST',
	headers: {
	  'Content-Type': 'application/json',
	  Accept: 'application/json',
	},
	params: { hapikey: bundle.authData.hapi_key },					// v1 auth for engagements API is different to v3 auth for CRM API. Yay.
	body: engagement
	};

	var ticket_ids = [];
	z.console.log('Tickets input value: ' + bundle.inputData.tickets);											
	for (const ticket of bundle.inputData.tickets)
	{	// Zapier forces us to use an AoH as input. Convert to A
		ticket_ids.push(ticket.id);
	}
	z.console.log('Tickets IDs array: '+ ticket_ids);											
	engagement.associations.ticketIds = ticket_ids;
	// Adjust the body content according to what happened.
	switch(bundle.inputData.event_type) {
		case "delivery_success":
			engagement.metadata.status = "COMPLETED";
			engagement.metadata.disposition = "f240bbac-87c9-4f6e-bf70-924b57d47db7";
			break;
		case "received":
			engagement.metadata.status = "COMPLETED";
			engagement.metadata.disposition = "f240bbac-87c9-4f6e-bf70-924b57d47db7";
			break;
		case "delivery_failure":
			engagement.metadata.status = "FAILED";
			break;
	}
	// Get the time the event ocurred
	var dateTime = new Date(bundle.inputData.time);
	engagement.engagement.timestamp = dateTime.getTime();
	z.console.log('Engagement: ' + JSON.stringify(engagement, null, 4));											

	// Make the call
	const response = await z.request(options);
	response.throwForStatus();
	const data = response.json;
	z.console.log('Response: ' + JSON.stringify(data, null, 4));											
	return { modified_count: bundle.inputData.tickets.length - data.associationCreateFailures.length };
};

module.exports = {
  operation: {
    perform: perform,
    inputFields: [
      {
        key: 'tickets',
        label: 'Tickets to update',
        children: [
			{
				key: 'id',
				type: 'integer',
				label: 'Ticket ID',
				required: true,
			}
        ],
        required: true,
        list: false,
        altersDynamicFields: false,

      },
      {
        key: 'message',
        label: 'The content of the message that was sent or received',
        type: 'string',
        required: true,
        altersDynamicFields: false,
      },
      {
        key: 'from_phone',
        label: 'Phone number that message was sent from',
        type: 'string',
        required: true,
        altersDynamicFields: false,
      },
      {
        key: 'to_phone',
        label: 'Phone number that message was sent to',
        type: 'string',
        required: true,
        altersDynamicFields: false,
      },
      {
        key: 'event_type',
        label: 'Event type',
        choices: { delivery_success:'SMS delivery succeeded', delivery_failure:'SMS delivery failed', received:'SMS received' },
        required: true,
        altersDynamicFields: false,
      },
      {
        key: 'time',
        label: 'Date and time the event occurred',
        type: 'datetime',
        required: true,
        altersDynamicFields: false,
      },
    ],
    sample: {
      modified_count: 2,
    },
    outputFields: [
      { key: 'modified_count', label: 'Count of tickets modified', type: 'integer' },
    ],
  },
  key: 'sms_event',
  noun: 'Event',
  display: {
    label: 'Add SMS Event to Tickets',
    description:
      'Add an event to the timeline of one or more tickets as a CALL engagement, describing a send or receive event of an SMS. For use with Sakari, which only logs events to contacts.',
    hidden: false,
    important: true,
  },
};


searches/find_tickets.js
const perform = async (z, bundle) => {
  const options = {
    url: 'https://api.hubapi.com/crm/v3/objects/tickets/search',
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Accept: 'application/json',
      Authorization: `Bearer ${bundle.authData.access_token}`,
    },
    body:
      '{"filterGroups":' +
      bundle.inputData.filter +
      ',"sorts":[{"propertyName":"createdate","direction":"DESCENDING"}],"limit": 100}',
  };

  return z.request(options).then((response) => {
    response.throwForStatus();
    const data = response.json;

	var filtered_results = [];
    // If we have a list of valid states, only include results that match them
	if(typeof(bundle.inputData.state_inclusion) != "undefined")
	{
			for (const result of data.results)
			{
				if(bundle.inputData.state_inclusion[result.properties.hs_pipeline_stage] == 1)
				{
					filtered_results.push(result);
				}
			}
	}
	else
	{	// pass all results through unfiltered
		filtered_results = data.results;
	}

	return [{ results: filtered_results, count: filtered_results.length }];
  });
};

module.exports = {
  operation: {
    perform: perform,
    inputFields: [
      {
        key: 'filter',
        label: 'Filter',
        type: 'text',
        helpText:
          'See the [HubSpot CRM API](https://developers.hubspot.com/docs/api/crm/tickets), "search" endpoint. Make sure your filter is enclosed in [ ]',
        required: true,
        list: false,
        altersDynamicFields: false,
      },
      {
        key: 'state_inclusion',
        label: 'Status inclusion',
        helpText:
          "For each possible value of hs_pipeline_stage, state whether a ticket should be included in the results. To include a ticket, set to 1. Otherwise set to 0. If the ticket's value for hs_pipeline_stage is not found in this dictionary, the default is to not include it (0). If your search does not filter on hs_pipeline, you will need to include values from all the possible pipelines to make sure all tickets get returned. A maximum of 100 tickets is returned by the HubSpot API BEFORE this filtering is carried out, so do not make your search too broad.\n\nIf this field is empty, all tickets are returned.",
        dict: true,
        required: false,
        altersDynamicFields: false,
      },
    ],
    sample: {
      id: '161521798',
      properties: {
        content: 'You can ignore this',
        createdate: '2020-07-16T13:50:11.754Z',
        hs_lastmodifieddate: '2021-11-16T15:27:27.612Z',
        hs_object_id: '161521798',
        hs_pipeline: '0',
        hs_pipeline_stage: '4',
        hs_ticket_category: null,
        hs_ticket_priority: 'LOW',
        subject: 'Test 2',
      },
      createdAt: '2020-07-16T13:50:11.754Z',
      updatedAt: '2021-11-16T15:27:27.612Z',
      archived: false,
    },
    outputFields: [
      { key: 'id', label: 'ID' },
      { key: 'properties__content', label: 'Description' },
      { key: 'properties__createdate', label: 'Create Date', type: 'datetime' },
      { key: 'properties__hs_pipeline', label: 'Pipeline' },
      { key: 'properties__hs_pipeline_stage', label: 'Status' },
      { key: 'subject', label: 'Title' },
    ],
  },
  key: 'find_tickets',
  noun: 'Ticket',
  display: {
    label: 'Find Tickets Advanced',
    description:
      'Finds tickets using a flexible JSON search filter. See the entry search under https://developers.hubspot.com/docs/api/crm/tickets for help buiding a filter. Maximum 100 results returned BEFORE any filtering, so do not make your search too broad. Tickets opened most recently will be first.',
    hidden: false,
    important: true,
  },
};
